sklearn/datasets/_openml.py

```diff
@@ -782,15 +896,50 @@ def fetch_openml(
             "unusable. Warning: {}".format(data_description["warning"])
         )
 
-    return_sparse = False
-    if data_description["format"].lower() == "sparse_arff":
-        return_sparse = True
+    valid_parser = ("auto", "pandas", "liac-arff")
+    if parser not in valid_parser:
+        raise ValueError(
+            f"`parser` must be one of {','.join(valid_parser)}. Got {parser} instead."
+        )
 
-    if as_frame == "auto":
-        as_frame = not return_sparse
+    if as_frame not in ("auto", True, False):
+        raise ValueError(
+            f"`as_frame` must be one of 'auto', True, or False. Got {as_frame} instead."
+        )
+
+    return_sparse = data_description["format"].lower() == "sparse_arff"
+    as_frame = not return_sparse if as_frame == "auto" else as_frame
+    if parser == "auto":
+        parser = "liac-arff" if return_sparse else "pandas"
+
+    if as_frame or parser == "pandas":
```

Comment:
Currently:

```python
mnist = fetch_openml("mnist_784", version=1, as_frame=False)
```

fails when pandas is not installed. I think we need to fall back to the liac arff parser?

Reply:
We discussed the issue here: https://github.com/scikit-learn/scikit-learn/pull/21938#discussion_r802521644

It mainly boils down to change of `dtypes` when switching parser.

Reply:
I see the reasoning. Currently the error for `fetch_openml("mnist_784", as_frame=False)` is:

```
ImportError: Using `parser='pandas'` requires pandas to be installed.
Alternatively, explicitely set `parser='liac-arff'`.
```

where the default is `parser="auto"`. Can we update the error to say `parser='auto'`?

I am concerned about is how this error breaks backward compatibility. `as_frame=False` works without pandas installed in version 1.0. If we want to stay backward compatible, we would need to set the default parser to `liac-arff` and go through a deprecation cycle to change the default.

Reply:
Fair point. +1 for raising a `FutureWarning` when `parser="auto"` and `as_frame=False` on dense datasets for now.